moved/src/move_execution/tests.rs

```diff
@@ -484,6 +515,107 @@ fn test_eoa_base_token_transfer() {
     assert_eq!(receiver_balance, transfer_amount);
 }
 
+#[test]
+fn test_treasury_charges_l1_cost_to_sender_account() {
+    // Initialize state
+    let genesis_config = GenesisConfig::default();
+    let mut signer = Signer::new(&PRIVATE_KEY);
+    let (_, mut state) = deploy_contract("natives", &mut signer, &genesis_config);
+
+    // Mint tokens in sender account
+    let sender = EVM_ADDRESS;
+    let mint_amount = U256::from(123u64);
+    let (tx_hash, tx) = create_deposit_transaction(mint_amount, sender);
+    let outcome =
+        execute_transaction(&tx, &tx_hash, state.resolver(), &genesis_config, 0, &()).unwrap();
+    state.apply(outcome.changes).unwrap();
+
+    let eth_treasury = AccountAddress::ZERO;
+    let base_token = MovedBaseTokenAccounts::new(eth_treasury);
+    let l1_cost = 1;
+
+    // Transfer to receiver account
+    let receiver = ALT_EVM_ADDRESS;
+    let transfer_amount = mint_amount.wrapping_shr(1);
+    let (tx_hash, tx) = create_transaction_with_value(
+        &mut signer,
+        TxKind::Call(receiver),
+        Vec::new(),
+        transfer_amount,
+    );
+    let outcome = execute_transaction(
+        &tx,
+        &tx_hash,
+        state.resolver(),
+        &genesis_config,
+        l1_cost,
+        &base_token,
+    )
+    .unwrap();
+    outcome.vm_outcome.unwrap();
+    state.apply(outcome.changes).unwrap();
+
+    let mint_amount = mint_amount.as_limbs()[0];
+    let transfer_amount = transfer_amount.as_limbs()[0];
+    let sender_balance = quick_get_eth_balance(&sender.to_move_address(), state.resolver());
+    let receiver_balance = quick_get_eth_balance(&receiver.to_move_address(), state.resolver());
+    let treasury_balance = quick_get_eth_balance(&eth_treasury, state.resolver());
+    assert_eq!(sender_balance, mint_amount - transfer_amount - l1_cost);
+    assert_eq!(receiver_balance, transfer_amount);
+    assert_eq!(treasury_balance, l1_cost);
+}
+
+#[test]
+fn test_treasury_does_not_charge_l1_cost_on_failure() {
```

Comment:
This actually isn't the right behavior. If a transaction fails then the user must still be charged for the gas and L1 cost because we still did the work to process the transaction and it will still be persisted on the L1. The only time we do not charge is when the transaction is invalid such that it was not executed at all in the first place (e.g. wrong signature, wrong nonce, not enough balance to cover gas/L1 fees, etc).

Reply:
My understanding was that if a transaction that the L2 receives fails, it won't be sent to L1 at all. I guess failed transactions are persisted on L1 too.